File: app/src/main/java/com/fsp/ticket_resto_app/ui/accueil/Accueil.java
package com.fsp.ticket_resto_app.ui.accueil;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.LauncherActivity;
import android.content.DialogInterface;
import android.content.Intent;
import android.icu.util.Currency;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResult;
import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.fsp.ticket_resto_app.PersistanceKt;
import com.fsp.ticket_resto_app.R;
import com.fsp.ticket_resto_app.controlleur.Controlleur;
import com.fsp.ticket_resto_app.ui.parametre.Parametre;
import com.fsp.ticket_resto_app.utilitaire.Popup;

import static com.fsp.ticket_resto_app.utilitaire.UtilitaireKt.simpleFormat;

import java.util.Locale;

import kotlin.Pair;

/**
 * Le controlleur de l'onget Accueil.
 *
 * @author François de Saint Palais
 */
public class Accueil extends AppCompatActivity {

    private static final String TAG = "Accueil";

    private Controlleur controlleur;

    private Button btnCalculer;
    private Button btnParametre;
    private EditText montantAPayer;
    private TextView labelPremierTicket;
    private TextView labelSecondTicket;
    private TextView quantitePremierTicket;
    private TextView quantiteSecondTicket;
    private TextView valeurReste;

    private ActivityResultLauncher<Intent> goParametreLauncher;

    private void initialiseComposants() {
        btnCalculer = findViewById(R.id.btnCalculer);
        btnParametre = findViewById(R.id.btnParametre);
        montantAPayer = findViewById(R.id.montantAPayer);
        labelPremierTicket = findViewById(R.id.labelPremierTicket);
        labelSecondTicket = findViewById(R.id.labelSecondTicket);
        quantitePremierTicket = findViewById(R.id.quantitePremierTicket);
        quantiteSecondTicket = findViewById(R.id.quantiteSecondTicket);
        valeurReste = findViewById(R.id.valeurReste);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        this.setContentView(R.layout.fragment_accueil);

        initialiseComposants();
        controlleur = Controlleur.getControlleur();
        goParametreLauncher = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(), this::retourActivity);

        btnCalculer.setOnClickListener(v -> calculer());
        btnParametre.setOnClickListener(v -> goParametre());

        loadPersistance();
        resetVue();
    }

    private void retourActivity(ActivityResult result) {
        resetVue();
    }

    private void calculer() {
        String textSaisie;
        textSaisie = montantAPayer.getText().toString();
        if (textSaisie.isEmpty()) {
            popupErreur("Veillez saisir un montant");
        } else {
            try {
                double montant = Double.parseDouble(textSaisie);
                controlleur.setMontantAPayer(montant);
                quantitePremierTicket.setText(controlleur.getQuantiteTicket1() + "");
                quantiteSecondTicket.setText(controlleur.getQuantiteTicket2() + "");
                valeurReste.setText(simpleFormat(controlleur.getReste()) + getMoneySymbol());
                Log.d("Accueil", "calculer: " + controlleur.getValeurTicket1() + " " + controlleur.getValeurTicket2() + " " + montant);
            } catch (NumberFormatException e) {
                popupErreur("Veillez saisir un entier");
            } catch (IllegalArgumentException e) {
                popupErreur("Vous devez saisir une valeur positive");
            }
        }
    }

    private static String getMoneySymbol() {
        return Currency.getInstance(Locale.getDefault()).getSymbol();
    }

    /**
     * Remet à zéro les champs de l'interface graphique.
     */
    private void resetVue() {
        labelPremierTicket.setText(String.format(getString(R.string.txt_label_quantite_ticket), simpleFormat(controlleur.getValeurTicket1())) + " " + getMoneySymbol());
        labelSecondTicket.setText(String.format(getString(R.string.txt_label_quantite_ticket), simpleFormat(controlleur.getValeurTicket2())) + " " + getMoneySymbol());
        quantitePremierTicket.setText("0");
        quantiteSecondTicket.setText("0");
        valeurReste.setText(0 + getMoneySymbol());
    }

    private void goParametre() {
        Intent intent = new Intent(this, Parametre.class);
        goParametreLauncher.launch(intent);
    }

    /**
     * Affiche une popup d'erreur.
     *
     * @param message le message à afficher.
     */
    private void popupErreur(String message) {
        Popup.Companion.showErreur(message, this);
    }

    private void toastErreur(String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    private void loadPersistance() {
        if (controlleur != null) {
            kotlin.Pair<Double, Double> valeurs = PersistanceKt.chargerValeurs(this);
            controlleur.setValeurTicket1(valeurs.component1());
            controlleur.setValeurTicket2(valeurs.component2());
        }
    }
}

File: app/src/main/java/com/fsp/ticket_resto_app/ui/accueil/Accueil.kt
package com.fsp.ticket_resto_app.ui.accueil

import android.annotation.SuppressLint
import android.content.ContentValues.TAG
import android.content.Intent
import android.icu.util.Currency
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.fsp.ticket_resto_app.R
import com.fsp.ticket_resto_app.chargerValeurs
import com.fsp.ticket_resto_app.controlleur.Controlleur
import com.fsp.ticket_resto_app.ui.parametre.Parametre
import com.fsp.ticket_resto_app.utilitaire.showErreur
import com.fsp.ticket_resto_app.utilitaire.simpleFormat
import java.util.Locale

/**
 * Le controlleur de l'onget Accueil.
 *
 * @author François de Saint Palais
 */
@SuppressLint("SetTextI18n")
class Accueil : AppCompatActivity() {

    private val TAG = "Accueil"

    private var btnCalculer: Button? = null
    private var btnParametre: Button? = null
    private var labelPremierTicket: TextView? = null
    private var labelSecondTicket: TextView? = null
    private var quantitePremierTicket: TextView? = null
    private var quantiteSecondTicket: TextView? = null
    private var valeurReste: TextView? = null
    private var montant: EditText? = null

    private val moneySymbol: String = Currency.getInstance(Locale.getDefault()).symbol

    private var goParametreLauncher: ActivityResultLauncher<Intent>? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.fragment_accueil)

        Log.d(TAG, "onCreate: Début")

        initialiseComposants()

        goParametreLauncher =
            registerForActivityResult<Intent, ActivityResult>(ActivityResultContracts.StartActivityForResult()) { result: ActivityResult ->
                this.retourActivity(result)
            }

        loadPersistance()
        resetVue()

        Log.d(TAG, "onCreate: Fin")
    }

    private fun initialiseComposants() {
        btnCalculer = findViewById(R.id.btnCalculer)
        btnParametre = findViewById(R.id.btnParametre)
        labelPremierTicket = findViewById(R.id.labelPremierTicket)
        labelSecondTicket = findViewById(R.id.labelSecondTicket)
        quantitePremierTicket = findViewById(R.id.quantitePremierTicket)
        quantiteSecondTicket = findViewById(R.id.quantiteSecondTicket)
        valeurReste = findViewById(R.id.valeurReste)
        montant = findViewById(R.id.montantAPayer)

        Log.d(TAG, btnCalculer.toString())
        Log.d(TAG, btnParametre.toString())
        btnCalculer!!.setOnClickListener {
            calculer()
            Log.d(this.TAG, "onCreate: Click Caluler")
        }
        btnParametre!!.setOnClickListener {
            goParametre()
            Log.d(this.TAG, "onCreate: Click Parametre")
        }
    }

    private fun retourActivity(result: ActivityResult) {
        resetVue()
    }

    private fun calculer() {
        val saisieMontant = montant?.text.toString()
        if (saisieMontant.isEmpty()) {
            showErreur("Veuillez saisir un montant", this)
        } else {
            try {
                val montant = saisieMontant.toDouble()
                Controlleur.setMontantAPayer(montant)
                quantitePremierTicket?.text = "${Controlleur.quantiteTicket1}"
                quantiteSecondTicket?.text = "${Controlleur.quantiteTicket2}"
                valeurReste?.text = simpleFormat(Controlleur.reste) + moneySymbol
            } catch (e: NumberFormatException) {
                showErreur("Le montant saisi n'est pas valide", this)
            }
        }

    }

    /**
     * Remet à zéro les champs de l'interface graphique.
     */
    private fun resetVue() {
        labelPremierTicket?.text = "${simpleFormat(Controlleur.valeurTicket1)}$moneySymbol"
        labelSecondTicket?.text = "${simpleFormat(Controlleur.valeurTicket2)}$moneySymbol"

        quantitePremierTicket?.text = "0"
        quantiteSecondTicket?.text = "0"

        valeurReste?.text = "0$moneySymbol"
    }

    private fun goParametre() {
        val intent = Intent(this, Parametre::class.java)
        goParametreLauncher?.launch(intent)
    }

    private fun loadPersistance() {
        val valeurs = chargerValeurs(this)
        Controlleur.valeurTicket1 = valeurs.component1()
        Controlleur.valeurTicket2 = valeurs.component2()
    }
}